### DISCUSSION TITLE
Using multiple clients in a NextJS and Express turborepo.

### DISCUSSION THREAD

**Discussion by ariburaco** [trpc/trpc#6061] [General]
How can I effectively use multiple tRPC clients in a Next.js and Express turborepo project? I'm trying to set up two separate tRPC clients - one for Next.js API routes and another for an Express server - but I'm encountering issues where only the inner provider seems to be working. 

I am using the "next" version trpc client and server packages.

What's the best approach to configure and use both clients simultaneously in this monorepo structure?

**Comment by Francowerner** [comment]
Did you find a solution? i've been working on it (we have the same exact goal) but i cant get it to work. Either the trpc context providers get all mixed up (i have separate instances of trpc to use in the frontend as well as context and queryClients) but trpc either hits the correct server on the routes that dont have the ExpressContext or it does when it shouldnt. I cant get it to consistently call different instances (one being the trpc server in packages/api and the other being my express trpc server in packages/express) without causing errors. And i didnt even mention CORS, which is turning it impossible to make it work. I've added al the headers everywhere that i possibly could, i've even deleted cors from express but it still returns the same error. And next-auth is a nightmare to make work, since its a very nuanced outcome i want.  I want to have create-t3-turbo working as is and add the functionality needed to have a separate package for a trpc express server that consumes the same session that the auth package provides (adapted to its needs, i did implement the authjs express adapter) and have a trpc client i can use inside of nextjs/expo that i can call depending on which of the apis i want to use. For long lived queries, events, and mutations etc i want to use the express trpc server and for everything else i want to use the serverless packages/api. Been working on it for 6 days, tried everything but it's just too confusing to make it work, each documentation tells you to do something different and there are a lot of edge cases because of the turborepo that render most docs useless. I've searched for examples that implement one or more of them to try and make them work myself but to no success sadly.

**Reply by KATT** [reply]
Please use linebreaks, this text is illegible

**Comment by ariburaco** [comment]
@Francowerner 

I made a workaround like this, maybe it can be useful:




````
UnifiedRouter = AppRouter & ExpressRouter;
export const api = createTRPCReact<UnifiedRouter>();

type ApiContext = {
  source: 'next' | 'express';
};

export function TRPCReactProvider(props: { children: React.ReactNode }) {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        refetchOnWindowFocus: false,
      },
    },
  });

  const [trpcClient] = useState(() =>
    api.createClient({
      links: [
        splitLink({
          condition: (op) => {
            const ctx = op.context as ApiContext;
            return ctx?.source === 'next';
          },
          true: unstable_httpBatchStreamLink({
            transformer: SuperJSON,
            url: `${APP_URL}/api/trpc`,
            async headers() {
              const headers = new Headers();
              // set additional headers if needed

              headers.set('x-trpc-source', 'react-next');
              return headers;
            },
          }),
          false: unstable_httpBatchStreamLink({
            transformer: SuperJSON,
            url: `${EXPRESS_URL}/api/trpc`,
            async headers() {
              const headers = new Headers();
              // set additional headers if needed
              headers.set('x-trpc-source', 'express');
              return headers;
            },
          }),
        }),
      ],
    })
  );

  return (
    <QueryClientProvider client={queryClient}>
      <api.Provider client={trpcClient} queryClient={queryClient}>
        {props.children}
      </api.Provider>
    </QueryClientProvider>
  );
}


## Usage:

- Next
const { data } = api.router.path.useQuery(undefined, {
  context: {
    source: "next",
  },
});

- Express
const { data } = api.router.path.useQuery(undefined, {
  context: {
    source: "express",
  },
});

## Key note: if you use the same router path for the queries, it is using only one of the clients, because of react-query cache.
````
